Framework/DataHandling/test/ReadMaterialTest.h

```diff
@@ -0,0 +1,271 @@
+      setMaterial.chemicalSymbol = EMPTY;
+      setMaterial.atomicNumber = 1;
+      setMaterial.massNumber = 1;
+      setMaterial.sampleNumberDensity = EMPTY_DOUBLE_VAL;
+      setMaterial.zParameter = 1;
+      setMaterial.unitCellVolume = EMPTY_DOUBLE_VAL;
+      setMaterial.sampleMassDensity = EMPTY_DOUBLE_VAL;
+      return setMaterial;
+    }
+    ();
+    auto result = ReadMaterial::validateInputs(params);
+    TS_ASSERT_EQUALS(result["UnitCellVolume"],
+                     "UnitCellVolume must be provided with ZParameter")
+  }
+
+  void testFailureValidateInputsSampleNumWithSampleMass() {
+    ReadMaterial::MaterialParameters params = [this]() -> auto {
+      ReadMaterial::MaterialParameters setMaterial;
+      setMaterial.chemicalSymbol = EMPTY;
+      setMaterial.atomicNumber = 1;
+      setMaterial.massNumber = 1;
+      setMaterial.sampleNumberDensity = 1;
+      setMaterial.zParameter = EMPTY_DOUBLE_VAL;
+      setMaterial.unitCellVolume = EMPTY_DOUBLE_VAL;
+      setMaterial.sampleMassDensity = 1;
+      return setMaterial;
+    }
+    ();
+    auto result = ReadMaterial::validateInputs(params);
+    TS_ASSERT_EQUALS(
+        result["SampleMassDensity"],
+        "Can not give SampleMassDensity with SampleNumberDensity set")
+  }
+
+  void testMaterialIsCorrect() {
+    ReadMaterial reader;
+    reader.setMaterial(FORMULA, 0, 0);
+    reader.setNumberDensity(EMPTY_DOUBLE_VAL, 1, EMPTY_DOUBLE_VAL,
+                            EMPTY_DOUBLE_VAL);
+    reader.setScatteringInfo(1, 2, 3, 4);
+    auto material = reader.buildMaterial();
+    compareMaterial(material, 1, 1, 2, 3, 4, FORMULA);
+  }
+
+  void testGenerateScatteringInfo() {
+    ReadMaterial reader;
+    reader.setMaterial(FORMULA, 0, 0);
+    reader.setNumberDensity(EMPTY_DOUBLE_VAL, 1, EMPTY_DOUBLE_VAL,
+                            EMPTY_DOUBLE_VAL);
+    auto material = reader.buildMaterial();
+    compareMaterial(material, 1, 0.0184000000, 5.0800000022, 5.0800000022,
+                    5.1000000044, FORMULA);
+  }
+
+  void testNoMaterialFailure() {
+    ReadMaterial reader;
+    reader.setMaterial(EMPTY, 0, 0);
+    reader.setNumberDensity(EMPTY_DOUBLE_VAL, 1, EMPTY_DOUBLE_VAL,
+                            EMPTY_DOUBLE_VAL);
+    reader.setScatteringInfo(EMPTY_DOUBLE_VAL, EMPTY_DOUBLE_VAL,
+                             EMPTY_DOUBLE_VAL, EMPTY_DOUBLE_VAL);
+    TS_ASSERT_THROWS(reader.buildMaterial(), std::runtime_error);
+  }
+
+private:
+  const double EMPTY_DOUBLE_VAL = 8.9884656743115785e+307;
+  const std::string EMPTY = "";
+  const std::string FORMULA = "V";
+
+  void compareMaterial(std::unique_ptr<Material> &material,
+                       const double numberDensity,
+                       const double coherentXSection,
+                       const double incoherentXSection,
+                       const double absorbXSection, const double totalXSection,
+                       const std::string &formula) {
```

Comment:
Maybe instead of taking 6 arguments for the material we should make a `Material` object by hand or using `MaterialBuilder` and then pass that as an arg. Only the `ReadMaterial` class is under test here so that's ok.